File: android/app/src/main/java/com/lightningkite/kwifttemplate/shared/views/SelectDemoVG.kt
package com.lightningkite.kwifttemplate.shared.views

import android.view.View
import com.lightningkite.kwift.actual.weak
import com.lightningkite.kwift.observables.actual.bind
import com.lightningkite.kwift.observables.actual.bindText
import com.lightningkite.kwift.observables.shared.ConstantObservableProperty
import com.lightningkite.kwift.observables.shared.ObservableStack
import com.lightningkite.kwift.shared.captureWeak
import com.lightningkite.kwift.views.actual.ViewDependency
import com.lightningkite.kwift.views.actual.onClick
import com.lightningkite.kwift.views.shared.EntryPoint
import com.lightningkite.kwift.views.shared.ViewGenerator
import com.lightningkite.kwifttemplate.layouts.RowTestXml
import com.lightningkite.kwifttemplate.layouts.SelectDemoXml

class SelectDemoVG(stack: ObservableStack<ViewGenerator>) : ViewGenerator(), EntryPoint {
    val stack: ObservableStack<ViewGenerator>? by weak(stack)
    override val title: String get() = "Select Demo"

    val options: List<ViewGenerator> = listOf(
        PongDemoVG(),
        FloatingHexagonsDemoVG(),
        MarginTestsVG(),
        MultipleDemoVG(),
        DateButtonDemoVG(),
        MapDemoVG(),
        LocationDemoVG(),
        LoadImageDemoVG(),
        ControlsDemoVG(),
        ExampleContentVG(),
        ViewPagerDemoVG(stack),
        SegmentedControlDemoVG(),
        SliderDemoVG(),
        WeekDemoVG(),
        DateRangeDemoVG(),
        LoginDemoVG(stack),
        DrawableDemoVG(),
        PreviewVG()
    )

    fun selectVG(viewGenerator: ViewGenerator){
        stack?.push(viewGenerator)
    }

    override fun generate(dependency: ViewDependency): View {
        val xml = SelectDemoXml()
        val view = xml.setup(dependency)

        xml.list.bind(
            data = ConstantObservableProperty(options),
            defaultValue = options.first(),
            makeView = { obs ->
                val xml = RowTestXml()
                val view = xml.setup(dependency)
                xml.label.bindText(obs) { it -> it.title }
                xml.button.onClick(captureWeak(this){ self -> self.selectVG(obs.value) })
                return@bind view
            }
        )

        return view
    }

    override val mainStack: ObservableStack<ViewGenerator>?
        get() = stack
}

File: ios/Kwift Template/com/lightningkite/kwifttemplate/shared/views/SelectDemoVG.swift
//Package: com.lightningkite.kwifttemplate.shared.views
//Converted using Kwift2

import Foundation
import Kwift



public class SelectDemoVG: ViewGenerator {
    
    
    public weak var stack: ObservableStack<ViewGenerator>? 
    override public var title: String {
        get {
            return "Select Demo"
        }
    }
    public var options: Array<ViewGenerator>
    
    public func selectVG(viewGenerator: ViewGenerator) -> Void {
        stack?.push(viewGenerator)
    }
    public func selectVG(_ viewGenerator: ViewGenerator) -> Void {
        return selectVG(viewGenerator: viewGenerator)
    }
    
    override public func generate(dependency: ViewDependency) -> View {
        var xml = SelectDemoXml()
        var view = xml.setup(dependency)
        xml.list.bind(data: ConstantObservableProperty(options), defaultValue: options.first(), makeView: { (obs) in 
            var xml = RowTestXml()
            var view = xml.setup(dependency)
            xml.label.bindText(obs) { (it) in 
                it.title
            }
            xml.button.onClick(captureWeak(self) { (self) in 
                self.selectVG(obs.value)
            })
            return view
        })
        return view
    }
    override public func generate(_ dependency: ViewDependency) -> View {
        return generate(dependency: dependency)
    }
    
    public init(stack: ObservableStack<ViewGenerator>) {
        self.stack = stack
        let options: Array<ViewGenerator> = [ControlsDemoVG(), ExampleContentVG(), ViewPagerDemoVG(stack), SegmentedControlDemoVG(), SliderDemoVG(), WeekDemoVG(), DateRangeDemoVG(), LoginDemoVG(stack), DrawableDemoVG(), PreviewVG()]
        self.options = options
        super.init()
    }
    convenience public init(_ stack: ObservableStack<ViewGenerator>) {
        self.init(stack: stack)
    }
}
 
